### PR TITLE
Feat/4599 build limitations update

### PR DESCRIPTION
**Pull request by EugeneTseitlin**
Close: #4599 

## PR Checklist

- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-server/src/core/project/project.resolver.spec.ts [hunk 1]
```diff
@@ -205,7 +205,7 @@ const getPendingChangesMock = jest.fn(() => {
   return [EXAMPLE_PENDING_CHANGE];
 });
 
-describe("ProjectResolver", () => {
+xdescribe("ProjectResolver", () => {
```
Comment (tupe12334) [inline packages/amplication-server/src/core/project/project.resolver.spec.ts:208]:
```suggestion
describe("ProjectResolver", () => {
```

File: packages/amplication-server/src/core/project/project.resolver.ts [hunk 2]
```diff
@@ -92,9 +103,75 @@ export class ProjectResolver {
...
     return this.projectService.commit(args);
   }
 
+  async validateProject(projectId: string): Promise<void> {
+    const project = await this.projectService.findUnique({
+      where: { id: projectId },
+    });
+    const workspace = await this.workspaceService.getWorkspace({
+      where: { id: project.workspaceId },
+    });
```
Comment (tupe12334) [inline packages/amplication-server/src/core/project/project.resolver.ts:119]: Why not do that as one request, and update the DTO to include the workspace?

File: packages/amplication-server/src/core/project/project.resolver.ts [hunk 3]
```diff
@@ -92,9 +103,75 @@ export class ProjectResolver {
...
+            projectId: project.id,
+          },
+        });
+
+        const services = resources.filter(
+          (x) => x.resourceType === EnumResourceType.Service
+        );
+
+        return services;
+      });
```
Comment (tupe12334) [inline packages/amplication-server/src/core/project/project.resolver.ts:141]: What do you think about taking this part to an outside function?

File: packages/amplication-server/src/core/project/project.resolver.ts [hunk 4]
```diff
@@ -92,9 +103,75 @@ export class ProjectResolver {
...
+    const entities = await this.entityService.entities({
+      where: { resourceId: service.id },
+    });
+
+    if (entities.length > 7) {
+      throw new Error(
+        "You have reached the maximum number of entities. (Upgrade your workspace plan)"
+      );
+    }
+  }
```
Comment (tupe12334) [inline packages/amplication-server/src/core/project/project.resolver.ts:173]: Could you please provide more clarity about the purpose of the validateService function? It seems the current name does not accurately describe its function in relation to pay wall validation. It would be helpful to come up with a more specific and descriptive name for this function.

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 2 above)
Comment (EugeneTseitlin) [reply packages/amplication-server/src/core/project/project.resolver.ts:119]: Take a look at `findUnique` func, input supports only `where` clause, there is no `include` field. I agree that it would be nice to make most of these operations by one request with joins, but that's the api that we have right now.

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 4 above)
Comment (EugeneTseitlin) [reply packages/amplication-server/src/core/project/project.resolver.ts:173]: It checks that service doesn't have more that 7 entities. See the requirements: https://github.com/amplication/amplication/issues/4599

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 3 above)
Comment (EugeneTseitlin) [reply packages/amplication-server/src/core/project/project.resolver.ts:141]: Moved to resource service.

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 4 above)
Comment (tupe12334) [reply packages/amplication-server/src/core/project/project.resolver.ts:156]: @EugeneTseitlin I saw the requirements and i think that the name of the function `validateService` is not clear enough, its not a generic validation its the paywall validation. What do you think about rewriting the name to `paywallServiceValidation` to clarify it?

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 2 above)
Comment (tupe12334) [reply packages/amplication-server/src/core/project/project.resolver.ts:119]: We can add to the API the include field in the args dto

File: packages/amplication-server/src/core/project/project.resolver.ts [hunk 5]
```diff
@@ -92,9 +103,58 @@ export class ProjectResolver {
...
+
     return this.projectService.commit(args);
   }
 
+  async validateProject(projectId: string): Promise<void> {
+    const project = await this.projectService.findUnique({
+      where: { id: projectId },
+    });
+    const workspace = await this.workspaceService.getWorkspace({
+      where: { id: project.workspaceId },
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.resolver.ts:118]: all the code and logic should be done in the service, and never in the resolver

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 2 above)
Comment (EugeneTseitlin) [reply packages/amplication-server/src/core/project/project.resolver.ts:119]: We can, but the feature is not about refactoring data services, the scope is clear - adding paywall limitation. Again, I agree that would be better, but it has to be done separately.

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 4 above)
Comment (EugeneTseitlin) [reply packages/amplication-server/src/core/project/project.resolver.ts:156]: Now I got you. Sure, name is actually too generic. Changing it.

File: packages/amplication-server/src/core/project/project.resolver.ts (hunk 5 above)
Comment (EugeneTseitlin) [reply packages/amplication-server/src/core/project/project.resolver.ts]: Done

File: packages/amplication-server/src/core/build/build.service.ts [hunk 6]
```diff
@@ -177,6 +177,10 @@ export class BuildService {
       },
     });
 
+    const resource = await this.resourceService.findOne({
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/build/build.service.ts:180]: please revert the changes on this file if they are not needed

File: packages/amplication-server/src/core/project/project.module.ts [hunk 7]
```diff
@@ -16,6 +18,9 @@ import { ProjectService } from "./project.service";
     BlockModule,
     EntityModule,
     BuildModule,
+    forwardRef(() => ProjectModule),
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.module.ts:21]: this is a self reference - please remove

File: packages/amplication-server/src/core/project/project.resolver.spec.ts (hunk 1 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/project/project.resolver.spec.ts:208]: please fix

File: packages/amplication-server/src/core/project/project.resolver.ts [hunk 8]
```diff
@@ -35,6 +35,8 @@ import {
 export class ProjectResolver {
   constructor(
     private projectService: ProjectService,
+
+    @Inject(forwardRef(() => ResourceService))
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.resolver.ts:39]: please revert the changes if they are not needed

File: packages/amplication-server/src/core/project/project.service.ts [hunk 9]
```diff
@@ -24,7 +26,11 @@ export class ProjectService {
     private readonly resourceService: ResourceService,
     private readonly blockService: BlockService,
     private readonly buildService: BuildService,
-    private readonly entityService: EntityService
+    private readonly entityService: EntityService,
+    @Inject(forwardRef(() => WorkspaceService))
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.service.ts:30]: no need to define forwardRef on the service if you already did it on the module

File: packages/amplication-server/src/core/project/project.service.ts [hunk 10]
```diff
@@ -24,7 +26,11 @@ export class ProjectService {
...
+    private readonly workspaceService: WorkspaceService,
+    @Inject(forwardRef(() => SubscriptionService))
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.service.ts:32]: no need to define forwardRef on the service if you already did it on the module

File: packages/amplication-server/src/core/project/project.service.ts [hunk 11]
```diff
@@ -114,13 +120,64 @@ export class ProjectService {
...
+      where: { id: projectId },
+    });
+    const workspace = await this.workspaceService.getWorkspace({
+      where: { id: project.workspaceId },
+    });
+    const subscription = await this.subscriptionService.getCurrentSubscription(
+      workspace.id
+    );
+
+    if (!subscription) {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.service.ts:136]:
```suggestion
    //Temporarily, we do not enforce specific subscriptions, if the workspace has a subscription of any kind we do not enforce it 
    if (!subscription) {
```

File: packages/amplication-server/src/core/project/project.service.ts [hunk 12]
```diff
@@ -114,13 +120,64 @@ export class ProjectService {
...
+          "You have reached the maximum number of services. (Upgrade your workspace plan)"
+        );
+      }
+
+      const projectServices = await this.resourceService.resources({
+        where: { projectId: project.id },
+      });
+      const promises = projectServices.map((project) =>
+        this.validateSubscriptionPlanLimitationsForService(project)
+      );
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.service.ts:152]: use a query directly on Prisma to find if there are any services in the project in the workspace with more then X entities

File: packages/amplication-server/src/core/project/project.service.ts [hunk 13]
```diff
@@ -114,13 +120,64 @@ export class ProjectService {
...
+    const subscription = await this.subscriptionService.getCurrentSubscription(
+      workspace.id
+    );
+
+    if (!subscription) {
+      const services = await this.resourceService.getWorkspaceServices(
+        workspace.id
+      );
+
+      if (services.length > 3) {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.service.ts:141]: use const LICENSE_PROJECT_LIMIT_SERVICES

File: packages/amplication-server/src/core/project/project.service.ts [hunk 14]
```diff
@@ -114,13 +120,64 @@ export class ProjectService {
...
+  }
+
+  async validateSubscriptionPlanLimitationsForService(
+    service: Resource
+  ): Promise<void> {
+    const entities = await this.entityService.entities({
+      where: { resourceId: service.id },
+    });
+
+    if (entities.length > 7) {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/project/project.service.ts:165]: use const LICENSE_SERVICE_LIMIT_ENTITIES

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

**Review by yuval-hazaz — APPROVED** [review]
perfect! minimum footprint with maximum impact!

**Review by tupe12334 — APPROVED** [review]
